### ISSUE TITLE
Error when only assigning one of two typevars to concrete types (but no error when assigning two or zero)

### ISSUE THREAD

**Issue by blakehawkins** [python/mypy#14078]
**Bug Report**

When writing a generic class with two typevars, mypy accepts classmethods that (1) assign values based on both typevars and (2) assigns values based on no typevars; but does not accept a classmethod that assigns one value based on a typevar and one value based on a concrete value.

https://stackoverflow.com/q/74417503/1669576

**To Reproduce**

```python
from typing import Generic, TypeVar

U = TypeVar('U')
V = TypeVar('V')
class Example(Generic[U, V]):
    def __init__(self, a: U, b: V):
        self._a = a
        self._b = b

    @classmethod
    def first(cls, val: U):
        return cls(val, None)
```

**Expected Behavior**

mypy should infer that None binds to V

**Actual Behavior**

mypy gives a seemingly spurious error `error: Argument 2 to "Example" has incompatible type "None"; expected "V"  [arg-type]`

**Your Environment**

```
name: test
channels:
  - conda-forge
  - defaults
dependencies:
  - _libgcc_mutex=0.1=conda_forge
  - _openmp_mutex=4.5=2_kmp_llvm
  - bzip2=1.0.8=h7b6447c_0
  - ca-certificates=2022.9.24=ha878542_0
  - certifi=2022.9.24=pyhd8ed1ab_0
  - ld_impl_linux-64=2.38=h1181459_1
  - libffi=3.3=he6710b0_2
  - libgcc-ng=12.2.0=h65d4601_19
  - libstdcxx-ng=11.2.0=h1234567_1
  - libuuid=1.41.5=h5eee18b_0
  - llvm-openmp=14.0.6=h9e868ea_0
  - mypy=0.990=py310h5764c6d_2
  - mypy_extensions=0.4.3=py310h06a4308_0
  - ncurses=6.3=h5eee18b_3
  - openssl=1.1.1s=h166bdaf_0
  - pip=22.2.2=py310h06a4308_0
  - psutil=5.9.0=py310h5eee18b_0
  - python=3.10.6=haa1d7c7_1
  - python_abi=3.10=2_cp310
  - readline=8.2=h5eee18b_0
  - setuptools=65.5.0=py310h06a4308_0
  - sqlite=3.39.3=h5082296_0
  - tk=8.6.12=h1ccaba5_0
  - toml=0.10.2=pyhd3eb1b0_0
  - tomli=2.0.1=pyhd8ed1ab_0
  - typing_extensions=4.3.0=py310h06a4308_0
  - tzdata=2022f=h04d1e81_0
  - wheel=0.37.1=pyhd3eb1b0_0
  - xz=5.2.6=h5eee18b_0
  - zlib=1.2.13=h5eee18b_0
prefix: /home/b/miniconda3/envs/test
```

- Mypy version used: 0.990
- Mypy command-line flags: (none)
- Mypy configuration options from `mypy.ini` (and other config files): (none)
- Python version used: 3.10.6

**Comment by joelberkeley** [comment]
I've replied to your SO qu. What if I call `type(Example(0, 0)).first(0)`? We can't infer `cls` as `type[Example[int, None]]` in that case.

**Comment by blakehawkins** [comment]
Hey @joelberkeley I took a look at your reply and I roughly follow the logical flow.

As a high level comment: it does intuitively feel like a contrived example to use `type()` to indirectly call a class method, right? (I acknowledge that this is a dynamic programming language)

For some follow-up questions -- and I am out of my depth here as a non-python programmer -- I do think there are some reasonable things that could be inferred here:

1. Do you expect `type(val)` to include its generic bindings, or do something more like "type erasure"?
2. Do you expect `Example[type].inferable_of_other_type()` to succeed or fail? Which types should it bind? Should it be allowed by a type checker?

For 1, as an example:

```
class Example1(Generic[T]):
    def __init__(self, t: T):
    self._t = t

my_int_example = Example(1)

print(my_int_example._t)

type(my_int_example)
```

Actual behavior: the `type` is just the class, not bound to any generics. This is expected -- typing hints are not used during runtime

For 2, as an example:

```
class Example1(Generic[T]):
    def __init__(self, t: T):
      self._t = t

    @classmethod
    def of(cls, t: T):
        return cls(t)

print(Example1[str].of(1)._t)
```

Actual behavior: prints `1`

It seems to me like it would be reasonable to assume that types are erased when calling `type(Example1.of("str"))` and that inference should then by contextually reset.

It also seems reasonable that a type checker would warn when there's an explicit mismatch in the case of `Example1[str].of(1)`

And finally, it seems _reasonable_ to infer types based _only_ on the outer `of` in the case of `type(Example1.of("str")).of(1)`

Let me know which steps seem unsound?

**Comment by blakehawkins** [comment]
None of these things are clear to me (I'm not a python expert):

- Should `Class[T].classmethod()` be typed as Class[T]?
- Should `type(class[T]).classmethod()` be typed as Class[T]?
- Should `Class[T].classmethod_inferred_as_U()` ever cause a runtime failure?

Given that the type parameters available on a class instance when used as `cls` in a classmethod have no impact on the classmethod's runtime, I think all the answers above should be "no". If anything, a linter should warn that T is discarded.
